Google.Solutions.Compute/Iap/IapTunnelingEndpoint.cs

```diff
@@ -0,0 +1,111 @@
+// specific language governing permissions and limitations
+// under the License.
+//
+
+using Google.Apis.Auth.OAuth2;
+using Google.Solutions.Compute.Net;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Solutions.Compute.Iap
+{
+    public class IapTunnelingEndpoint : ISshRelayEndpoint
+    {
+        private const string BaseUri = "wss://tunnel.cloudproxy.app/v4/";
+        private const string SubprotocolName = "relay.tunnel.cloudproxy.app";
+        private const string Origin = "bot:iap-tunneler";
+        public const string DefaultNetworkInterface = "nic0";
+
+        // Cf. https://developers.google.com/identity/protocols/googlescopes#iapv1
+        public const string RequiredScope = "https://www.googleapis.com/auth/cloud-platform";
+
+        public VmInstanceReference VmInstance { get; }
+
+        public ushort Port { get; }
+
+        public string Interface { get; }
+
+        private readonly ICredential credential;
+
+        public IapTunnelingEndpoint(ICredential credential, VmInstanceReference vmInstance, ushort port, string nic)
+        {
+            this.credential = credential;
+            this.VmInstance = vmInstance;
+            this.Port = port;
+            this.Interface = nic;
+        }
+
+        public Task<INetworkStream> ConnectAsync(CancellationToken token)
+        {
+            return ReconnectAsync(null, 0, token);
+        }
+
+        public async Task<INetworkStream> ReconnectAsync(
+            string sid,
+            ulong ack,
+            CancellationToken token)
+        {
+            var urlParams = new Dictionary<string, string>
+            {
+                { "project", this.VmInstance.ProjectId },
+                { "zone", this.VmInstance.Zone },
+                { "instance", this.VmInstance.InstanceName },
+                { "interface", this.Interface },
+                { "port", this.Port.ToString() },
+                { "_", Guid.NewGuid().ToString() }  // Cache buster.
+            };
+
+            if (sid != null)
+            {
+                urlParams["sid"] = sid;
+                urlParams["ack"] = ack.ToString();
+            }
+
+            var queryString = string.Join(
+                "&",
+                urlParams.Select(kvp => kvp.Key + "=" + WebUtility.UrlEncode(kvp.Value)));
+
+            var uri = new Uri(
+                BaseUri + 
+                (sid == null ? "connect" : "reconnect" ) +
+                "?" +
+                queryString);
+
+            var accessToken = await this.credential.GetAccessTokenForRequestAsync(
+                null,
+                token).ConfigureAwait(false);
+
+            var websocket = new ClientWebSocket();
+            websocket.Options.AddSubProtocol(SubprotocolName);
+            websocket.Options.SetRequestHeader("Authorization", "Bearer " + accessToken);
```

Comment:
once the tunnel is up it stays up even if the unerlying connect authentication path as an expired access_token, yea/

i.,e the accessToken will expire in an hour but the tunnel will stay up and the token is never used again?

Reply:
Not sure how IAP behaves here -- it might actually close the connection. But the next time, we'll get a `CloseCode.NOT_AUTHORIZED` for sure.